### ISSUE TITLE
Fuzzy patient search

### ISSUE THREAD

**Issue by clemens-msupply** [openmsupply/open-msupply#502]
Return patients:
- with similar names
- with similar name and similar DoB ?

Idea:
- Narrow down server search result by looking for patient that match the first two chars of the first/last name.
- Use fuzzy search lib to further reduce the search results

**Comment by nisha-dangol** [comment]
Hi @clemens-msupply @roxy-dao @mark-prins 
Is this issue about how the search doesn't work properly when search-term contains the name; only when works when the search-term is exactly the patient-name? 
Recording reference: https://recordit.co/SyTStjh8nf (See that it searches for FN: August but not for FN: Aug)

Also, what does the download button mean if it fetches the patient but that patient doesn't show up at all after downloading, another issue?
<img width="1017" alt="image" src="https://github.com/msupply-foundation/open-msupply/assets/59193490/079e0e35-894c-488a-a491-de82abc96e5c">

**Comment by clemens-msupply** [comment]
I would say the existing patient search works as expected here. We don't have any fuzzy search so we do an exact match. You previously enter the patient details, and then we check if there is an existing patient. So what can happen here is that you made a spelling mistake, and then you need fuzzy matching to find similar patients...

**Comment by nisha-dangol** [comment]
> EDIT: Even though we could still do a LIKE query, for example, if you have previously entered a full version of a name "Alexander" and then the next time you just enter the short version "Alex".
>
I say, searching with short version term `Alex` should show up the existing patient `Alexander`. Not returning in search-result is a bug!

**Comment by raviSussol** [comment]
I tried to look at the flow in both o/msupply apps and found that this issue should more likely to be created in the desktop msupply repo as the change needs to be implemented there. Desktop [issue](https://github.com/msupply-foundation/msupply/issues/14497)

**Comment by clemens-msupply** [comment]
Since you can search for a patient on the local site and on the central server, both the remote site and the central server should ideally implement the same search algorithm.

**Comment by raviSussol** [comment]
> Since you can search for a patient on the local site and on the central server, both the remote site and the central server should ideally implement the same search algorithm.

Right! Would still be easier to implement same search logic as per oms in the central server (msupply side). https://github.com/msupply-foundation/msupply/issues/14515#issuecomment-2014495982

**Comment by raviSussol** [comment]
The only thing that needs implementing in the oms side is to add pagination in the search result datatable so that all result can be viewed IMO? Something like,
![Screenshot 2024-03-22 at 1 11 20 PM](https://github.com/msupply-foundation/open-msupply/assets/13145522/28039589-8e47-490e-9f10-5b70e4c35415)

**Comment by clemens-msupply** [comment]
omSupply doesn't support fuzzy search at the moment, e.g. if you search for "Allex" it doesn't give you results for "Alex".

Not sure if pagination is needed, e.g. if you search for first and last name and dob there shouldn't hopefully too many results. However, there might be cases where there is along list of matches, in this case pagination might be useful.

**Comment by raviSussol** [comment]
> omSupply doesn't support fuzzy search at the moment, e.g. if you search for "Allex" it doesn't give you results for "Alex".

Ah right! Thanks @clemens-msupply for clarifying it. Looks like I seemed to be confused between "fuzzy" and "like" search that is working in the oms atm.

> if you search for first and last name and dob there shouldn't hopefully too many results.

Oms still allows you to search by just first/last name field. And in that case, result of "fuzzy/like" search would generate more than 20 (which is default limit size in the msupply atm) from the central server.

**Comment by andreievg** [comment]
Triage, setting priority as nice to have and removing milestone for now.
